README.md
# ReCalc v1.4.0

Calculadora con funciones simples.

## Dependencias

- node > 18

## Comandos

```bash
# iniciar/limpiar la db
npm run seed

# iniciar calculadora
npm run start

# correr los test
npm run test

# correr test end-to-end
npm run test:e2e

# correr test estatico
npm run lint
```
## Integrantes

- Nicolas Morales

- Tomas Novelino

- Bautista Basilio

- Franco Borsella



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.4.0] 26-07-2023

## [1.3.3] 11-06-2023

### Added

- Se realizaron las conexiones del front con la api.
- Se le subio en nivel a las reglas del eslint.
- Se agregaron test e2e

### Fixed

- Se resolvieron errores marcados por test estatico

## [1.3.2] 08-06-2023

### Fixed

- Path donde corre eslint

### Fixed

- Problema al correr tests e2e en Windows

## [1.3.1] 05-06-2023

### Added

- Dependencia cross-env para correr tests e2e

### Fixed

- Problema al correr tests e2e en Windows

## [1.3.0] 05-06-2023

### Added

- Interfaz básica de la calculadora
- ESLint para tests estáticos
- Playwright para tests e2e

## [1.2.0] 15-05-2023

### Added

- Sequelize para el manejo de base de datos
- Tests de integración sobre API y modelo
- Integración continua

## [1.1.0] 20-04-2023

### Added

- Endpoint para la función de resta
- Dependencias para realizar tests
- Tests para la función de resta

## [1.0.0] 17-04-2023

### Added

- Interface CLI para realizar cálculos en forma interactiva
- Base de la API
- Funciones core para realizar suma, resta, multiplicación, división y potencia

[unreleased]: https://github.com/frlp-utn-ingsoft/recalc/compare/v1.4.0...HEAD
[1.4.0]: https://github.com/frlp-utn-ingsoft/recalc/releases/tag/v1.4.0
[1.3.3]: https://github.com/frlp-utn-ingsoft/recalc/releases/tag/v1.3.3
[1.3.2]: https://github.com/frlp-utn-ingsoft/recalc/releases/tag/v1.3.2
[1.3.1]: https://github.com/frlp-utn-ingsoft/recalc/releases/tag/v1.3.1
[1.3.0]: https://github.com/frlp-utn-ingsoft/recalc/releases/tag/v1.3.0
[1.2.0]: https://github.com/frlp-utn-ingsoft/recalc/releases/tag/v1.2.0
[1.1.0]: https://github.com/frlp-utn-ingsoft/recalc/releases/tag/v1.1.0
[1.0.0]: https://github.com/frlp-utn-ingsoft/recalc/releases/tag/v1.0.0


test/ui.test.js
import { test, expect } from '@playwright/test';
import { seed } from '../src/seed.js';
import { Operation, History } from '../src/models.js'

test.describe('test', () => {
  test.describe.configure({ mode: 'serial' });

  test.beforeEach(async () => {
    await seed();
  })

  test('Deberia tener como titulo de pagina recalc', async ({ page }) => {
    await page.goto('./');

    // Expect a title "to contain" a substring.
    await expect(page).toHaveTitle(/recalc/i);
  });

  test('Deberia poder realizar una resta', async ({ page }) => {
    await page.goto('./');

    await page.getByRole('button', { name: '7' }).click()
    await page.getByRole('button', { name: '9' }).click()
    await page.getByRole('button', { name: '-' }).click()
    await page.getByRole('button', { name: '9' }).click()

    const [response] = await Promise.all([
      page.waitForResponse((r) => r.url().includes('/api/v1/sub/')),
      page.getByRole('button', { name: '=' }).click()
    ]);

    const { result } = await response.json();
    expect(result).toBe(70);

    await expect(page.getByTestId('display')).toHaveValue(/70/)

    const operation = await Operation.findOne({
      where: {
        name: "SUB"
      }
    });

    const historyEntry = await History.findOne({
      where: { OperationId: operation.id }
    })

    expect(historyEntry.firstArg).toEqual(79)
    expect(historyEntry.secondArg).toEqual(9)
    expect(historyEntry.result).toEqual(70)
  });

  test('Deberia poder realizar una multiplicacion', async ({ page }) => {
    await page.goto('./');

    await page.getByRole('button', { name: '5' }).click()
    await page.getByRole('button', { name: '*' }).click()
    await page.getByRole('button', { name: '5' }).click()

    const [response] = await Promise.all([
      page.waitForResponse((r) => r.url().includes('/api/v1/multi/')),
      page.getByRole('button', { name: '=' }).click()
    ]);

    const { result } = await response.json();
    expect(result).toBe(25);

    await expect(page.getByTestId('display')).toHaveValue(/25/)

    const operation = await Operation.findOne({
      where: {
        name: "MUL"
      }
    });

    const historyEntry = await History.findOne({
      where: { OperationId: operation.id }
    })

    expect(historyEntry.firstArg).toEqual(5)
    expect(historyEntry.secondArg).toEqual(5)
    expect(historyEntry.result).toEqual(25)
  });

  test('Deberia poder realizar una suma', async ({ page }) => {
    await page.goto('./');

    await page.getByRole('button', { name: '8' }).click()
    await page.getByRole('button', { name: '+' }).click()
    await page.getByRole('button', { name: '1' }).click()
    await page.getByRole('button', { name: '3' }).click()

    const [response] = await Promise.all([
      page.waitForResponse((r) => r.url().includes('/api/v1/add/')),
      page.getByRole('button', { name: '=' }).click()
    ]);

    const { result } = await response.json();
    expect(result).toBe(21);

    await expect(page.getByTestId('display')).toHaveValue(/21/)

    const operation = await Operation.findOne({
      where: {
        name: "ADD"
      }
    });

    const historyEntry = await History.findOne({
      where: { OperationId: operation.id }
    })

    expect(historyEntry.firstArg).toEqual(8)
    expect(historyEntry.secondArg).toEqual(13)
    expect(historyEntry.result).toEqual(21)
  });

  test('Deberia poder realizar una división', async ({ page }) => {
    await page.goto('./');
  
    await page.getByRole('button', { name: '8' }).click()
    await page.getByRole('button', { name: '/' }).click()
    await page.getByRole('button', { name: '4' }).click()
  
    const [response] = await Promise.all([
      page.waitForResponse((r) => r.url().includes('/api/v1/div/')),
      page.getByRole('button', { name: '=' }).click()
    ]);
  
    const { result } = await response.json();
    expect(result).toBe(2);
  
    await expect(page.getByTestId('display')).toHaveValue(/2/)
  
    const operation = await Operation.findOne({
      where: {
        name: "DIV"
      }
    });
  
    const historyEntry = await History.findOne({
      where: { OperationId: operation.id }
    })
  
    expect(historyEntry.firstArg).toEqual(8)
    expect(historyEntry.secondArg).toEqual(4)
    expect(historyEntry.result).toEqual(2)
  });

  test('Deberia dar error al dividir por cero', async ({ page }) => {
    await page.goto('./');
  
    await page.getByRole('button', { name: '8' }).click()
    await page.getByRole('button', { name: '/' }).click()
    await page.getByRole('button', { name: '0' }).click()
  
    const [response] = await Promise.all([
      page.waitForResponse((r) => r.url().includes('/api/v1/div/')),
      page.getByRole('button', { name: '=' }).click()
    ]);
  
    const result = await response.json();
    expect(result.error).toBe("Division por cero.");
  
    await expect(page.getByTestId('display')).toHaveValue("Error: Division por cero.")
  });
  
  test('Deberia poder realizar la operacion potencia',async({ page})=>{
    await page.goto('./');
  
    await page.getByRole('button', { name: '5' }).click()
    await page.getByRole('button', { name: '^2' }).click()
  
    const [response] = await Promise.all([
      page.waitForResponse((r) => r.url().includes('/api/v1/pow/')),
      page.getByRole('button', { name: '=' }).click()
    ]);
  
    const { result } = await response.json();
    expect(result).toBe(25);
  
    await expect(page.getByTestId('display')).toHaveValue(/25/)
  
    const operation = await Operation.findOne({
      where: {
        name: "POW"
      }
    });
  
    const historyEntry = await History.findOne({
      where: { OperationId: operation.id }
    })
  
      expect(historyEntry.firstArg).toEqual(5)
      expect(historyEntry.result).toEqual(25)
  });

  test('Deberia dar error al ingresar un numero mayor a 100000', async ({ page }) => {
    await page.goto('./');

    await page.getByRole('button', { name: '4' }).click()
    await page.getByRole('button', { name: '0' }).click()
    await page.getByRole('button', { name: '0' }).click()
    
    await page.getByRole('button', { name: '0' }).click()
    await page.getByRole('button', { name: '0' }).click()
    await page.getByRole('button', { name: '0' }).click()
    await page.getByRole('button', { name: '^2' }).click()

    await page.getByRole('button', { name: '=' }).click()

    await expect(page.getByTestId('display')).toHaveValue("Error: Numero mayor a 100000")
    
  });

  test('Deberia traer la operacion 4+4=8 y 5-4=1 luego de haberlas realizada', async ({ page }) => {
    await page.goto('./');

    await page.getByRole('button', { name: '4' }).click()
    await page.getByRole('button', { name: '+' }).click()
    await page.getByRole('button', { name: '4' }).click()
    await page.getByRole('button', { name: '=' }).click()

    await page.getByRole('button', { name: '5' }).click()
    await page.getByRole('button', { name: '-' }).click()
    await page.getByRole('button', { name: '4' }).click()
    await page.getByRole('button', { name: '=' }).click()

    await page.getByRole('button', { name: 'Ver historial' }).click()

    await Promise.all([
      page.waitForResponse((r) => r.url().includes('/api/v1/histories')),
      page.getByRole('button', { name: 'Ver historial' }).click()
    ]);

    await expect(page.getByTestId('histories')).toHaveValue("4 + 4 = 8\n5 - 4 = 1\n")
    
  });

  test('Deberia borrar el historial luego de realizar una operacion', async ({ page }) => {
    await page.goto('./');

    await page.getByRole('button', { name: '4' }).click()
    await page.getByRole('button', { name: '+' }).click()
    await page.getByRole('button', { name: '4' }).click()
    await page.getByRole('button', { name: '=' }).click()

    page.on('dialog', dialog => dialog.accept())
    await page.getByRole('button', { name: 'Borrar historial' }).click()
    

    await Promise.all([
      page.waitForResponse((r) => r.url().includes('/api/v1/delete/all')),
      page.getByRole('button', { name: 'Borrar historial' }).click()
    ]);

    const histories = await History.findAll({
      where: {}
    })

    await expect(page.getByTestId('histories')).toHaveValue("")
    expect(histories.length).toBe(0)
    
  });


  test('Deberia poder borrar el display',async({ page})=>{
    await page.goto('./');
  
    await page.getByRole('button', { name: '7' }).click()
    await page.getByRole('button', { name: '6' }).click()
    await page.getByRole('button', { name: '+' }).click()
  
    await page.getByRole('button', { name: 'c' }).click()
    
    await expect(page.getByTestId('display')).toHaveValue("")
  
  });

  test('Deberia poder realizar la conversión de decimal a binario',async({ page})=>{
    await page.goto('./');
  
    await page.getByRole('button', { name: '4' }).click()
    await page.getByRole('button', { name: 'bin' }).click()
  
    const [response] = await Promise.all([
      page.waitForResponse((r) => r.url().includes('/api/v1/bin/')),
      page.getByRole('button', { name: '=' }).click()
    ]);
  
    const { result } = await response.json();
    expect(result).toBe("100");
  
    await expect(page.getByTestId('display')).toHaveValue(/100/)
  
    const operation = await Operation.findOne({
      where: {
        name: "BIN"
      }
    });

    const historyEntry = await History.findOne({
      where: { OperationId: operation.id }
    })
  
      expect(historyEntry.firstArg).toEqual(4)
      expect(historyEntry.result).toEqual(100)
  });

  test('Deberia poder realizar la raiz cuadrada',async({ page})=>{
    await page.goto('./');
  
    await page.getByRole('button', { name: '3' }).click()
    await page.getByRole('button', { name: '6' }).click()
    await page.getByRole('button', { name: 'sqrt' }).click()
  
    const [response] = await Promise.all([
      page.waitForResponse((r) => r.url().includes('/api/v1/sqrt/')),
      page.getByRole('button', { name: '=' }).click()
    ]);
  
    const { result } = await response.json();
    expect(result).toBe(6);
  
    await expect(page.getByTestId('display')).toHaveValue(/6/)
  
    const operation = await Operation.findOne({
      where: {
        name: "SQRT"
      }
    });
  
    const historyEntry = await History.findOne({
      where: { OperationId: operation.id }
    })
  
      expect(historyEntry.firstArg).toEqual(36)
      expect(historyEntry.result).toEqual(6)
  });

  test('Deberia poder realizar una suma por teclado',async({ page})=>{
    await page.goto('./');
  
    await page.keyboard.press('Digit4');
    await page.keyboard.press('NumpadAdd');
    await page.keyboard.press('Digit2');
  
    const [response] = await Promise.all([
      page.waitForResponse((r) => r.url().includes('/api/v1/add/')),
      page.keyboard.press('Enter')
    ]);
  
    const { result } = await response.json();
    expect(result).toBe(6);
  
    await expect(page.getByTestId('display')).toHaveValue(/6/)
  
    const operation = await Operation.findOne({
      where: {
        name: "ADD"
      }
    });
  
    const historyEntry = await History.findOne({
      where: { OperationId: operation.id }
    })
  
      expect(historyEntry.firstArg).toEqual(4)
      expect(historyEntry.secondArg).toEqual(2)
      expect(historyEntry.result).toEqual(6)
  });

  test('no deberia mostrar undefined en el display cuando se hace click en el "=" sin que halla una operacion',async({page})=>{
    await page.goto('./');

    await page.getByRole('button', { name: '=' }).click()

    await expect(page.getByTestId('display')).not.toHaveValue(/undefined/)
  })


})




test/api.test.js
const request = require('supertest');
const api = require('../src/api.js');
const { seed } = require('../src/seed.js')

beforeEach(async () => {
    await seed()
})


describe("API div", () => {
    test("Deberia responder con un 400 y si su segundo parametro es 0 devuelva mensaje de error", async () => {
        const app = await api.build();
        const mensajeError="Division por cero.";

        const res = await request(app).get("/api/v1/div/1/0")
        expect(res.body.error).toBe(mensajeError);
        expect(res.status).toBe(400);
    })
})

describe("API pow",()=>{
    test("Deberia responder con un error:400", async()=>{
        const app = await api.build();

        const res = await request(app).get('/api/v1/pow/y')
            

        expect(isNaN(res.body.result)).toEqual(true)
        expect(res.body.error).toBe("El parámetro no es un número")
        expect(res.status).toBe(400)
    
    })
    test("Deberia responder con un error:400", async()=>{

        const app = await api.build();

        const res = await request(app).get('/api/v1/pow/XC')
            

        expect(isNaN(res.body.result)).toEqual(true)
        expect(res.body.error).toBe("El parámetro no es un número")
        expect(res.status).toBe(400)
    
    })
    


})

describe("API add decimal",()=>{

    test("Debería responder 0.3 al sumar 0.1+0.2",async()=>{
        const app = await api.build()

        const res = await request(app).get('/api/v1/add/0.1/0.2')

        expect(isNaN(res.body.result)).toEqual(false)
        expect(res.error.text).toBe(undefined)
        expect(res.status).toBe(200)
        expect(parseFloat(res.body.result.toFixed(1))).toEqual(0.3)
    })
})


describe("API add", () => {
    test("Deberia responder con un 200 ok y su respuesta ser menor al primer parametro", async () => {
        const app = await api.build()
        const a = 4
        const b = -10

        const res = await request(app).get(`/api/v1/add/${a}/${b}`)


        expect(res.body.result).toBeLessThan(a)
        expect(res.headers['content-type']).toBe("application/json; charset=utf-8")
        expect(res.status).toBe(200)
    })

    test("Deberia responder con un 200 ok y su respuesta ser menor al primer parametro", async () => {
        const app = await api.build()
        const a = 0
        const b = -100000

        const res = await request(app).get(`/api/v1/add/${a}/${b}`)


        expect(res.body.result).toBeLessThan(a)
        expect(res.headers['content-type']).toBe("application/json; charset=utf-8")
        expect(res.status).toBe(200)
    })
})

describe("API history", () => {
    test('Debería responder con un 200 ok y verificar si la entrada del historial si existe', async () => {
        const app = await api.build();
        //Agrego una entrada al historial para asegurarme que siempre haya al menos una con id 1
        await request(app).get('/api/v1/add/2/2');
        const response = await request(app).get(`/api/v1/history/1`);

        expect(response.statusCode).toBe(200);
        expect(response.body.data.id).toBe(1);
    });

    test('Deberia devolver un error si la entrada del historial no existe', async () => {
        const app = await api.build();
        const response = await request(app).get('/api/v1/history/-1');

        expect(response.statusCode).toBe(404);
        expect(response.body.error).toBe('Entrada del historial no encontrada')
    })
})

describe ("API multi",()=>{

    function isFloat(num){
        return num % 1 !== 0
    }

    test("Deberia responder con un 200 ok", async () => {
        const app = await api.build()
        const a = 5.5
        const b = 2.4


        const res = await request(app).get(`/api/v1/multi/${a}/${b}`)
        expect(res.status).toBe(200)
        
        expect(isFloat(res.body.result)).toBe(true)
    })

    
})

describe("API sqrt",()=>{
    test("Deberia responder con un 200 ",async()=>{
        const app = await api.build();

        const a = 64;

        const res = await request(app).get(`/api/v1/sqrt/${a}`)

        expect(res.status).toBe(200)
        expect(isNaN(res.body.a)).toBe(true)
        expect(res.body.result).toBe(8)

    })

    test("Deberia responder con un 200 ",async()=>{
        const app = await api.build();

        const a = 36;

        const res = await request(app).get(`/api/v1/sqrt/${a}`)

        expect(res.status).toBe(200)
        expect(isNaN(res.body.a)).toBe(true)
        expect(res.body.result).toBe(6)
    })
})


describe ("API get all histories",()=>{

    test("Deberia responder con un 200 y deberia devolver 3 entidades.", async () => {
        const app = await api.build()

        await request(app).get(`/api/v1/add/1/2`)
        await request(app).get(`/api/v1/add/1/4`)
        await request(app).get(`/api/v1/sub/1/2`)

        const res = await request(app).get(`/api/v1/histories`)
        
        expect(res.status).toBe(200)
        expect(res.body.allHistories.length).toBe(3)
    })

    test("Deberia responder con un 200 y deberia devolver 2 entidades pasandole el filtro de ADD.", async () => {
        const app = await api.build()
        
        await request(app).get(`/api/v1/add/1/2`)
        await request(app).get(`/api/v1/add/1/4`)
        await request(app).get(`/api/v1/sub/1/2`)

        const res = await request(app).get(`/api/v1/histories?operation=ADD`)
        
        expect(res.status).toBe(200)
        expect(res.body.allHistories.length).toBe(2)
    })

    test("Deberia responder con un 200 y deberia devolver 2 entidades pasandole como parametro page 2 y size 2", async () => {
        const app = await api.build()
        
        await request(app).get(`/api/v1/add/1/2`)
        await request(app).get(`/api/v1/add/1/4`)
        await request(app).get(`/api/v1/sub/1/2`)
        await request(app).get(`/api/v1/sub/1/2`)

        const res = await request(app).get(`/api/v1/histories?page=2&size=2`)
        
        expect(res.status).toBe(200)
        expect(res.body.allHistories.length).toBe(2)
    })

    test("Deberia responder con un 200 y deberia devolver 1 entidades pasandole como parametro operation ADD, page 2 y size 2", async () => {
        const app = await api.build()
        
        await request(app).get(`/api/v1/add/1/2`)
        await request(app).get(`/api/v1/add/1/4`)  
        await request(app).get(`/api/v1/sub/1/2`)
        await request(app).get(`/api/v1/div/1/2`)
        await request(app).get(`/api/v1/add/1/4`)

        const res = await request(app).get(`/api/v1/histories?operation=ADD&page=2&size=2`)
        
        expect(res.status).toBe(200)
        expect(res.body.allHistories.length).toBe(1)
    })
})

describe("API delete/all",()=>{
    test("Deberia responder un 200 y eliminar todas las filas",async ()=>{
        const app = await api.build()
        
        await request(app).get(`/api/v1/add/1/2`)
        await request(app).get(`/api/v1/add/1/3`)  
        await request(app).get(`/api/v1/sub/1/4`)

        const res = await request(app).get(`/api/v1/delete/all`)

        expect(res.status).toBe(200)
        expect(res.body.message).toBe("Se borraron todas las historias de la tabla")
        
    })
})

test/core.test.js
import core from '../src/core.js'
import cli from '../src/cli.js'

describe('Subtract', () => {
    test('Deberia 2 - 2 = 0', () => {
        expect(core.sub(2, 2)).toBe(0);
    })

    test('Deberia 6 - 4 = 2', () => {
        expect(core.sub(6, 4)).toBe(2);
    })
})

describe('Divide', () => {
    test('Deberia 6 / 2 = 3', () => {
    
        const a = 6;
        const b = 2;

        const result = core.div(a, b);

        expect(result).toBe(3);
    })

    test('No deberia 10 / 2 = 4', () => {

        const a = 10;
        const b = 2;

        const result = core.div(a, b)

        expect(result).not.toBe(4)
    })

    //Opcional

    test('Deberia 10 / 0 = Error: Division por cero', () => {

        const a = 10;
        const b = 0;

        const result = () => {return core.div(a, b)}

        expect(result).toThrow(new Error("Division por cero."));
    })

})

describe('Pow', () => {
    test('Deberia 5 ^ 2 = 25', () => {
        
        const a=5;
        
        const result = core.pow (a);
        expect(result).toBe(25); 
    })

    test('No Deberia 2 ^ 2 = 10', () => {
        
        const a=2;

        const result = core.pow (a);
        expect(result).not.toBe(10); 
    })

    test('Deberia -4 ^ 2 = 16', () => {
        
        const a=-4;
        
        const result = core.pow (a);
        expect(result).toBe(16); 
    })
})

describe('Multiply',()=>{
    test('Deberia 3 * 3 = 9', ()=>{
        const a = 3;
        const b = 3;

        const result = core.mul(a,b);

        expect(result).toBe(9);
    })
    test('No deberia 3 * 2 = 12',()=>{
        const a = 3;
        const b = 2;

        const result = core.mul(a,b);

        expect(result).not.toBe(12);
    })

    test('Deberia -4 * 3 = -12', ()=>{
        const a = -4;
        const b = 3;
    
        const result = core.mul(a,b);
    
        expect(result).toBe(-12);
    })
})

describe('Add', () => {
    test('Deberia 2 + 2 = 4', () => {
        //Given
        const a = 2
        const b = 2

        //When
        const result = core.add(a, b)

        //Then
        expect(result).toBe(4)
    })

    test('No deberia 1 + 2 = 4', () => {
        //Given
        const a = 1
        const b = 2

        //When
        const result = core.add(a, b)
    })
})

describe('decimalToBinary', () => {
    test('Deberia convertir 10 a 1010', () => {
        const a = 10;

        const result = core.bin(a);
        expect(result).toBe('1010');
    })

    test('Deberia convertir 42 a 101010', () => {
        const a = 42;

        const result = core.bin(a);
        expect(result).toBe('101010');
    })

    test('N debería convertir 20 a 11011', () => {
        const a = 20;

        const result = core.bin(a);
        expect(result).not.toBe('11011');
    })
})


describe('cli loop func test', () => {
    
    test('Deberia retornar 👋👋👋 si se ingresa exit', async () =>{
        //Given
        const mockLogFunction = jest.fn()
        const mockReadLine = {
            question: jest.fn(),
            close: jest.fn()
        }
        mockReadLine.question.mockReturnValueOnce('exit')

        //When
        const result = await cli.loop(mockReadLine, Object.keys(core), mockLogFunction)

        //Then
        expect(mockLogFunction).toHaveBeenCalledWith("👋👋👋")
    })
    
    test('Deberia retornar "Funcion invalida, intente nuevamente" si se ingresa una funcion no valida', async () => {
        //Given
        const mockLogFunction = jest.fn()
        const mockReadLine = {
            question: jest.fn().mockReturnValueOnce("funcionInvalida").mockReturnValueOnce("exit"),
            close: jest.fn()
        }

        //When
        await cli.loop(mockReadLine, Object.keys(core), mockLogFunction);
      
        //Then
        expect(mockLogFunction).toHaveBeenCalledWith("Funcion invalida, intente nuevamente");
    })
})


describe('test: 2do parametro mayor que el primero',()=>{
    test('Deberia 1 - 2 = -1', () => {
    
        const a=1;
        const b=2;
        
        const result = core.sub(a, b);
        expect(result).toBe(-1);
    
    
    })    
})


describe('cli throw func test',()=>{
    test('Deberia retornar un error en caso de no ser un numero ',()=>{
        
        function isNum(){
            cli.isNumber("a");
        }

        expect(isNum).toThrow(new Error("El valor ingresado no es un numero"));

    })
})





src/cli.js
import { createInterface } from "node:readline/promises";
import core from "./core.js"

const readlineFunction = createInterface({
    input: process.stdin,
    output: process.stdout,
});

const ALL_FNS_NAMES = Object.keys(core);

const AVAILABLE_FNS= [...ALL_FNS_NAMES, 'exit'].join(', ')

function isNumber(num){
    if(isNaN(num)){
        throw new Error("El valor ingresado no es un numero")
    }
        
    return num;
}

async function loop(readline, allFnsNames, logFunction = console.log) {
    const fnName = await readline.question(`Ingrese función (${AVAILABLE_FNS}): `)
    let firstNum;
    let secondNum;

    if (fnName === "exit") {
        logFunction("👋👋👋");
        return readline.close();
    }

    if(!allFnsNames.includes(fnName)){
        logFunction("Funcion invalida, intente nuevamente");
        return loop(readline, allFnsNames);
    }

    const fn = core[fnName];

    try{
        firstNum = isNumber(await readline.question("Ingrese el primer número: ")) 
        secondNum = fnName ==="pow" ? null : isNumber(await readline.question("Ingrese el segundo número: "))
   
    }catch(error){
        logFunction(error.messagegit)
        return loop(readline, allFnsNames);
    } 

    const result = fnName === "pow"? fn(Number(firstNum)) : fn(Number(firstNum), Number(secondNum));

    logFunction(result);
    return loop(readline, allFnsNames);
}

loop(readlineFunction, ALL_FNS_NAMES);


export default {
                loop:loop, 
                isNumber:isNumber
            }

src/models.js
import { Sequelize, DataTypes } from 'sequelize';

var storageDB;

if(process.env.NODE_ENV === 'test'){
    storageDB = ":memory:"
}else{
    storageDB = './db.sqlite3'
}

const inTest = process.env.NODE_ENV === 'test' || process.env.NODE_ENV === 'test:e2e';

const sequelize = new Sequelize({
    dialect: 'sqlite',
    logging: !inTest,
    storage: storageDB
})

export const History = sequelize.define('History', {
    firstArg: {
        type: DataTypes.NUMBER,
        allowNull: true
    },
    secondArg: {
        type: DataTypes.NUMBER,
        allowNull: true
    },
    result: {
        type: DataTypes.NUMBER,
        allowNull: true
    },
    error: {
        type: DataTypes.TEXT,
        allowNull: true
    }
});

export const Operation = sequelize.define('Operation', {
    name: {
        type: DataTypes.STRING,
        allowNull: false
    },
});

Operation.hasMany(History)
History.belongsTo(Operation)

export async function createHistoryEntry({ firstArg, secondArg, operationName, result, error }) {
    const operation = await Operation.findOne({
        where: {
            name: operationName
        }
    });

    return History.create({
        firstArg,
        secondArg,
        result,
        OperationId: operation.id,
        error
    })
}


export async function allHistory(filter, page = 1, size = 10){

    const includeOptions = [{
        model: Operation
    }];

    if (filter !== undefined) {
        includeOptions[0].where = { name: filter };
    }

    const histories = await History.findAll({
        include: includeOptions,
        limit: size,
        offset: (page - 1) * size
    });
    
    return histories; 
}

export async function deleteHistory(filter, page = 1, size = 10){
    
    const histories = await allHistory(filter, page, size)

    histories.forEach(async history => {
        await history.destroy({
            where: {}
        });
    })
}

export function createTables() {
    return Promise.all([
        History.sync({ force: true }),
        Operation.sync({ force: true })
    ]);
}



export function findByID(id) {
    return History.findByPk(id);
}

src/public/index.js
const $display = document.querySelector('.display')
const $hisotriesDisplay = document.querySelector('.histories-display')
const $buttons = document.querySelector('.buttons')
const $hisotries = document.querySelector('.hisotries')

const operations = ['-','+','^2','*', '/','bin','sqrt'];

const operationsMap = new Map([    
[1,'+'],
[2,'-'],    
[3,'*'],
[4,'/'],
[5,'**'],
[6,'sqrt'],
[7,'bin']
]);

let currentDisplay = "";
let operation = null;
let reset = false;

var $toggle = document.getElementById('container')
var $body = document.querySelector('body')

$toggle.onclick= function(){
    $toggle.classList.toggle('active');
    $body.classList.toggle('active');
}

$hisotries.addEventListener('click', async (e) => {
    let result;

    const nextAction = e.target.name
    if(nextAction === "getAllHistory"){
        result = await getAllHistory()
    }

    if(nextAction === "deleteAllHistory"){
        const confirmResult = confirm("¿Seguro que quiere borrar el historial?")
        if(confirmResult){
            result = await deleteAllHistory()
            result = ""
        }else{
            return;
        }
    }

    if(nextAction === "clearScreen"){
        result = "";
    }

    return renderHistoriesDisplay(result)
})


document.addEventListener('keydown', async (e) => {
    const key = e.key;
    const buttonId = getButtonIdForKey(key);
    if (buttonId) {
      const button = document.getElementById(buttonId);
      if (button) {
        button.click();
      }
    }
  });

$buttons.addEventListener('click', async (e) => {
    var nextAction = "";

    if (e.target.name !== "c" && e.target.name !== undefined){
        nextAction = e.target.name
    }

    if (nextAction === "=") {
        const [firstArg, secondArg] = currentDisplay.split(operation)

        let result;

        if (operation === "-") {
            result = await calculateSub(firstArg, secondArg)
        }

        if (operation === "+") {
            result = await calculateAdd(firstArg, secondArg)
        }
        
        if (operation === "*"){
            result = await calculateMult (firstArg, secondArg)
        }
        if (operation === "^2"){
            result =  await calculatePow(firstArg)
        }

        if (operation === "/") {
            result = await calculateDiv(firstArg, secondArg)
        }

        if (operation === "bin"){
            result =  await calculateBin(firstArg)
        }
        if (operation === "sqrt"){
            result = await calculateSqrt(firstArg)
        }

        if(result !== undefined){
            reset = true;
            return renderDisplay(result);
        }
    }

    if (e.target.name === "c"){
        renderDisplay("")
    }

    if (operations.includes(nextAction)) {
        operation = nextAction;
    }

    if (reset) {
        reset = false;
        operation = null;

        renderDisplay(nextAction);
    } else {

        if(nextAction !== "=") {
            let chars = currentDisplay + nextAction
            if( !isNaN(chars[0])){
                renderDisplay(currentDisplay + nextAction);
            }
        }

    }
})

async function calculateSub(firstArg, secondArg) {
    const resp = await fetch(`/api/v1/sub/${firstArg}/${secondArg}`)
    const { result } = await resp.json();

    return result;
}

async function calculateMult(firstArg,secondArg) {
    const resp = await fetch(`/api/v1/multi/${firstArg}/${secondArg}`)
    const {result} = await resp.json();

    return result;
}

async function calculatePow(firstArg){
    if(firstArg > 100000){
            return "Error: Numero mayor a 100000"
    }
    const resp = await fetch(`/api/v1/pow/${firstArg}`)
    const {result} = await resp.json();
    
    return result;
}

async function calculateAdd(firstArg, secondArg){
    const resp = await fetch(`/api/v1/add/${firstArg}/${secondArg}`)
    const { result } = await resp.json();

    return result;
}

async function calculateDiv(firstArg, secondArg){
    const resp = await fetch(`/api/v1/div/${firstArg}/${secondArg}`);
    if(resp.status === 200){
        const { result } = await resp.json();
        return result;
    }
    if(resp.status === 400){
        const result  = await resp.json();
        return "Error: " + result.error;
    }
}

async function getAllHistory(){
    const resp = await fetch(`/api/v1/histories`)
    const result = await resp.json();
    return result.allHistories;
}

async function deleteAllHistory(){
    const resp = await fetch(`/api/v1/delete/all`)
    const result = await resp.json();
    return result;
}

async function calculateBin(firstArg){
    const resp = await fetch(`/api/v1/bin/${firstArg}`)
    const {result} = await resp.json();
    
    return result;
}

async function calculateSqrt(firstArg){
    const resp = await fetch (`/api/v1/sqrt/${firstArg}`)
    const {result} = await resp.json();

    return result;
}


function getButtonIdForKey(key) {
    const keyMap = {
      '0': 'button-0',
      '1': 'button-1',
      '2': 'button-2',
      '3': 'button-3',
      '4': 'button-4',
      '5': 'button-5',
      '6': 'button-6',
      '7': 'button-7',
      '8': 'button-8',
      '9': 'button-9',
      '.': 'button-.',
      '+': 'button-+',
      '-': 'button--',
      '*': 'button-*',
      '/': 'button-/',
      'Enter': 'button-=',
      'Backspace': 'button-c',
      'b': 'button-bin',
      's': 'button-sqrt',
      'p': 'button-^2'
    };
  
    return keyMap[key] || '';
  }

function renderDisplay(chars) {
    currentDisplay = chars;
    $display.value = chars;
    
}


function renderHistoriesDisplay(histories){
    $hisotriesDisplay.value = ""
    if(histories === ""){
        return;
    }
    histories.forEach(history => {
        $hisotriesDisplay.value += history.firstArg + " " + operationsMap.get(history.OperationId) + " " + history.secondArg + " = " + history.result + "\n"
    });
   
}

